### DISCUSSION TITLE
How to get the indices of the variable for the Khalil2016() observation.

### DISCUSSION THREAD

**Discussion by kevin031060** [ds4dm/ecole#223] [Questions]
I'm confused about the shape of the Khalil2016() observation.
At a solving step of a instance, I found the shape of the Khalil2016() observation is (146, 72), the shape of the action_set is 59, and the shape of NodeBipartite().edge_features is (270, 1000).
As the action_set is the candidates set, why there are 146 set of features for the Khalil2016() observation. It should be (59, 72)?. If not, what are the indices of the variable for the Khalil2016() observation.

**Comment by AntoinePrv** [comment]
The action set of the `Branching` environment returns the branching candidates variables by default (otherwise pseudo branching candidates)
https://github.com/ds4dm/ecole/blob/05485f225987c7a4ba128d1ee6b5ca904dccd103/libecole/src/dynamics/branching.cpp#L21

`Khalil2016` returns the features for all pseudo branching candidates variables.
https://github.com/ds4dm/ecole/blob/05485f225987c7a4ba128d1ee6b5ca904dccd103/libecole/src/observation/khalil-2016.cpp#L578

Finally, `NodeBipartite` is computed for all LP columns.
https://github.com/ds4dm/ecole/blob/05485f225987c7a4ba128d1ee6b5ca904dccd103/libecole/src/observation/nodebipartite.cpp#L174

`NodeBipartite` by definition extract all the variables to fully represent the current node.
Pseudo branching candidates variables are a superset of the branching candidates.
I think we could add a flag to `Khalil2016` to avoid computing this superset when it is not needed (by default).

**Reply by kevin031060** [reply]
Thank you! It helps a lot for me. But I didn't find the PseudoBranchCands() function in the Pyscipopt doc. There is only the getLPBranchCands(). In SCIP doc, there is a SCIPgetPseudoBranchCands() function. So should I modify the code of Pyscipopt or ecole and compile them to get the indices for the Khalil2016() obsersvation?

**Reply by AntoinePrv** [reply]
I think you're right. I think we're missing a vector that would give us the variable ID for all lines.
Let me get back to you

**Reply by kevin031060** [reply]
I'm now working on a paper using the ecole. Thanks a lot!

**Reply by AntoinePrv** [reply]
We're missing some additional vector in the observation function to get for each variable an ID to match with the action set.
We'll add this in the release. In the meantime, I don't think you can use it properly.

**Reply by kevin031060** [reply]
Thank you.

**Reply by gasse** [reply]
Indeed, in its current state the observation function is unfortunately barely usable, due to the points @AntoinePrv mentioned.

I just wanted to add that I think you should still be able to use the Khalil2016 features if you also ask for the pseudo branching candidates in the branching environment (which can be specified as an optional parameter when the environment is instantiated). Then the Khalil2016 features should be exactly the size of the branching candidate vector, and the variables should follow the same ordering, hence you should be able to use them directly with a 1 to 1 correspondance with the action set. I haven't actually tried it myself, so this remains to be confirmed still.

Best,
Maxime
